jax/experimental/jax2tf/jax2tf.py

```diff
@@ -126,6 +127,119 @@ def add_unit(v: TfValOrUnit, aval: core.AbstractValue):
     return core.unit if aval is core.abstract_unit else v
   return util.safe_map(add_unit, vals, avals)
 
+
+class SingletonDecorator:
+  def __init__(self, cls):
+    self.cls = cls
+    self.instance = None
+
+  def __call__(self, *args, **kwargs):
+    if not self.instance:
+      self.instance = self.cls(*args, **kwargs)
+    return self.instance
+
+PrimitiveName = str
+ErrorType = str
+ErrorString = str
+Devices = Tuple[str, ...]
+Limitation = Tuple[PrimitiveName, ErrorType, ErrorString, Devices]
+
+def categorize(prim: core.Primitive, *args: TfVal, **kwargs) \
+    -> List[Limitation]:
+  """
+  Given a primitive and a set of parameters one would like to pass to it,
+  categorize identifies the potential limitations the call would encounter when
+  converted to TF through jax2tf.
+
+  Args:
+    prim: the primitive to call.
+    args: the arguments to pass to prim.
+    kwargs: the keyword arguments to pass to prim.
+
+  Returns:
+    A list of limitations
+  """
+  limitations: List[Limitation] = []
+  all_devices = ["CPU", "GPU", "TPU"]
+
+  def _report_failure(error_type: str, msg: str,
+                      devs: Sequence[str] = all_devices) -> None:
+    limitations.append((prim.name, error_type, msg, tuple(devs)))
+
+  tf_unimpl = functools.partial(_report_failure, "Missing TF support")
+
+  if prim in [lax.min_p, lax.max_p]:
+    tf_dtype = args[0].dtype
```

Comment:
This suggests that args are `TfVal`. This is fine for use in tests, but not for use in classifying usage. In that case the arguments are JAX values. That is why I had suggested that perhaps `np.shape` and `np.dtype` works for all kinds of values that we care about.

Reply:
The way this is executed, the args are indeed `TfVal` for now; maybe I should make it explicit in the type signature?

Reply:
But then I cannot use this function to count usage in JAX. If I use it there, the args are JAX tracers. That is Ok, as long as we have a way to get the shape and dtype that works for both.

Reply:
FYI, `np.dtype(dtype)` fails if `dtype` is a TensorFlow dtype (one needs to get `dtype.as_numpy_dtype` first, or better, `to_jax_dtype(dtype)`).
I think I addressed your comment in my latest commit (?).